### ISSUE TITLE
Make a survey of languages with syntactic macros

### ISSUE THREAD

**Issue by masak** [masak/007#489]
Starting from the languages listed in https://en.wikipedia.org/wiki/Macro_(computer_science)#Syntactic_macros, make a table with the languages on the rows and capabilities on the columns.

After the table, also summarize each language in prose, pointing out tradeoffs and limitations; details that the table is too coarse to contain. Include plenty of links to resources on the web such as reference documentation and papers.

(Do not assume Wikipedia's list of languages is complete; if there are other languages that seem to fit the description of having syntactic macros, include them. For example, I'm pretty sure we also want to have SweetJS in there.)

**Comment by masak** [comment]
The intent here being, of course, to explore the prior art in the field, both in order to be able to relate 007 to other things in some kind of ontology, and in order to <del>loot and pillage</del> find inspiration from other languages, especially those which have walked the same path and faced the same foes.

**Comment by vendethiel** [comment]
I see sweetjs in the Wiki article (though perhaps you added it?).
I’m also sad it’s now defunct.

What are the points we’re most interested aboutin other impls? :)

**Comment by claes-magnus** [comment]
Looking forward to seeing this table! 👍

**Comment by masak** [comment]
Just want to link this one in here: [Dylan and the Lisp family tree's central core](https://axisofeval.blogspot.com/2010/04/dylan-and-lisp-family-trees-central.html). The list of languages in that post seems worth following up on:

* Common Lisp
* Dylan
* EuLisp
* ISLISP
* Goo
* PLOT

**Comment by masak** [comment]
It seems I might have to include [Wat](http://axisofeval.blogspot.com/2013/05/a-new-low-in-programming-language.html) in my list, too.

**Comment by masak** [comment]
Linking this [syntax-rules primer](http://web.archive.org/web/20060616054033/home.comcast.net/~prunesquallor/macro.txt) from here, just to have it linked from somewhere. It also relates somewhat to #491.

**Comment by vendethiel** [comment]
> You cannot write an `infix' macro, nor can you write a macro in a
`nested position', i.e., an expression like ((foo x) y z) is
always considered a procedure application.  (The subexpression (foo x)
will be rewritten of course, but it will not be able to affect the
processing of subexpressions y and z.)  This will be important later
on.

The first part is untrue, for us. The second part is at opposites with some other features 007 wants.

**Comment by masak** [comment]
[Kernel](https://web.cs.wpi.edu/~jshutt/kernel.html).

**Comment by masak** [comment]
And also [Virtua](http://www.manuelsimoni.net/virtua/index.html), by the Wat author (and "the Axis of Eval" blogger).

Needless to say, some of these languages (especially the fexpr ones) are not mainstream, and I won't be able to write a lot about them. But there will still be something to say about "how does the language do macro-ish things?".

**Comment by masak** [comment]
> I see sweetjs in the Wiki[pedia] article (though perhaps you added it?).

No, but I think I might have missed it, or forgot it was there.

> I’m also sad it’s now defunct.

Any idea why?

> What are the points we’re most interested aboutin other impls? :)

For the table: approach to macro hygiene/gensyms, etc. Type of macros (procedural/pattern-matching).

For the prose, some details about how the macro system "feels", what limits it sets up for itself wrt syntax and semantics, maybe what known weaknesses it has (such as Common Lisp and accidental capture), but balance it out with what things it provides (such as anaphoric macros).

**Comment by vendethiel** [comment]
> > I’m also sad it’s now defunct.
> 
> Any idea why?

Probably a collection of factors: Disnet used to work at Mozilla, and they might've had time for open-source work (after all, sweetjs used to be under the mozilla github organisation). Maybe just lack of interest.
Nate Faubion, early contributor + big part of the ecosystem (made the modules for ADT, pattern matching, etc) has joined the PureScript movement.
As for the latest contributor, I don't actually know. Maybe a simple lack of time is the answer.

**Comment by masak** [comment]
Also perhaps include [Z](https://chrisdone.com/posts/z/) in the survey, although it seems a bit <del>buggy</del> awesome.

[HN discussion](https://news.ycombinator.com/item?id=20557363).

**Comment by masak** [comment]
Also throwing [this HN discussion](https://news.ycombinator.com/item?id=20605603) in here.

I feel I need to get somewhat of a handle on `syntax-rules`, `syntax-case`, and `syntax-parse`.

**Comment by masak** [comment]
I think [`Filter::Inject`](https://github.com/LanX/Filter-Inject) deserves to go on this list also.

**Comment by vendethiel** [comment]
The use of a string reminds me of [BlackCoffee](https://github.com/paiq/blackcoffee) somewhat.

**Comment by masak** [comment]
Also [macropy](https://github.com/lihaoyi/macropy).

**Comment by masak** [comment]
[MetaOCaml](http://okmij.org/ftp/ML/MetaOCaml.html). To the extent that I've looked at it, it has quasiquotation and unquotes, and these interact with the type system.

**Comment by masak** [comment]
Maybe [D](https://dlang.org/) is worth mentioning in this context, because their official stance seems to be that they ended up [not needing macros](https://dlang.org/articles/pretod.html).

> D doesn't have a preprocessor; but D provides a more scalable means to solve the same problems.

Granted, it's preprocessor/text macros they're talking about here, not real AST macros. But still.

I think the most important part of that page is the last one, about [template mixins](https://dlang.org/spec/template-mixin.html). Haven't looked deeply, but it seems that they do merit the inclusion of D in this list.

**Comment by masak** [comment]
Scala has [def macros](https://docs.scala-lang.org/overviews/macros/overview.html) (still experiemental) and deserves at place in the survey.

Rust also [has macros](https://doc.rust-lang.org/1.29.0/book/2018-edition/appendix-04-macros.html), and the way to define them reminds me of `syntax-rules` a bit.

**Comment by masak** [comment]
[The T reference manual](http://web.archive.org/web/20170515045004/http://repository.readscheme.org/ftp/papers/t_manual.pdf) (and by the way, I don't know yet if T deserves a slot of its own in this survey) contains this interesting tidbit: "T's quasiquote facility, inherited from the Maclisp family of Lisp...".

Maclisp goes back a long, long way. It originated in the late 60s, and influenced Interlisp and later Common Lisp. But the claim that quasiquotes come from that language family is not backed up by the [Maclisp reference manual](http://www.softwarepreservation.org/projects/LISP/MIT/Moon-MACLISP_Reference_Manual-Apr_08_1974.pdf), Revision Ø, from 1974. Maybe it was added later? Or there were variants/dialects of Maclisp, and quasiquotes weren't in the main one?

**Comment by masak** [comment]
Nor does this [Interlisp reference manual](http://bitsavers.trailing-edge.com/pdf/xerox/interlisp/Interlisp_Reference_Manual_1974.pdf) (also from 1974) mention quasiquotes.

**Comment by vendethiel** [comment]
Not sure if Scala def macros should « count » still, since they were deprecated for Scala 3/Dotty.

**Comment by masak** [comment]
Wow, a Big Rewrite. Where have I heard that before? 😄 

Anyway, unless I've missed a big announcement, it's not a question of "they were deprecated", and more a question of "they [wioll haven be](http://pages.cs.wisc.edu/~param/quotes/guide.html) deprecated". We'll cross that Subinverted Plagal Past Subjunctive when we get to it.

**Comment by masak** [comment]
As to how/where quasiquotes originated, [The History of Lisp](https://dreamsongs.com/Files/HOPL2-Uncut.pdf) has much of the story. (See pp72-75.)

Of note, they were originally called "pseudoquotes". (That is, the relation to Quine's [quasiquotation](https://en.wikipedia.org/wiki/Quasi-quotation) is not there from the start.) One of the early practitioners of the art, including of nested quasiquotes, was Alan Bawden. Searching for "alan bawden mit nested quasiquote" gives [this .ps.gz](http://people.csail.mit.edu/alan/ftp/quasiquote-v59.ps.gz), which seems on skimming to be a very nice read. ([PDF version](https://3e8.org/pub/scheme/doc/Quasiquotation%20in%20Lisp%20(Bawden).pdf).)

On the way I also found [this 1993 email](https://groups.csail.mit.edu/mac/ftpdir/scheme-mail/HTML/rrrs-1993/msg00064.html) where Jonathan Rees recounts Alan Bawden opining strongly that quasiquote expansion should be a part of the _reader_. (I don't know about other implementations, but it isn't in Bel; it's a macro. That is, you can write either \``(a ,b)` or `(bquote (a (comma b))`, and they result in the same macro-expanded expression.)

[This text](https://pdfs.semanticscholar.org/c698/c99c5e732da6ea484c4b8804a92a69f7d5c0.pdf) mirthily quotes from Bawden's article, as well as provides a nice example of double-unquote, a `defun` macro:

```
(defmacro defun (name params &rest body)
    `(defmacro ,name ,params
        `((lambda ,',params ,',@body)
        ,,@params)
    )
)
```

(Edit: I had too many backquotes; two of them should be regular quotes. Now fixed.)

I use this macro especially because something tells me @vendethiel used it as an example years ago, and now he'll smack me over the head with it. Either for being a slow learner, or for having a memory like a colander, or both.

**Comment by masak** [comment]
A later reference work about MacLisp is apparently called "The Revised MACLISP Manual" but nicknamed [The Pitmanual](http://www.maclisp.info/pitmanual/index.html) after its author, Kent M. Pitman.

This one [does mention backquoting and splicing](http://www.maclisp.info/pitmanual/syntax.html). So all in all it would seem that quasiquoting and the backquote syntax was indeed invented during the evolution of the MacLisp family of LISPs.

**Comment by masak** [comment]
> something tells me @vendethiel used it as an example years ago

Ah yes, here: https://github.com/masak/alma/issues/30#issuecomment-225585820 &mdash; 3.5 years ago.

**Comment by vendethiel** [comment]
> and now he'll smack me over the head with it. Either for being a slow learner, or for having a memory like a colander, or both.

I don't think I do either, if anything, I'm always surprised at how much you remember was mentioned around. :-)

**Comment by masak** [comment]
I guess I am my own harshest judge. 🙂 It's taken me this many years to wrap my head around all this stuff. To the extent I understand it even now, that is.

**Comment by masak** [comment]
[Groovy has macros](https://groovy-lang.org/releasenotes/groovy-2.5.html). Apparently, they grew out of their AST Transformers, and are documented together with them [here](https://groovy-lang.org/metaprogramming.html#_macros). From what I can see, they don't address hygiene at all. (I might be wrong.) Groovy's `macro` keyword is for quoting code, and seems to correspond to Alma's `quasi` keyword.

**Comment by masak** [comment]
@ALANVF Agreed, after checking the documentation of all those three. Hum, somehow I was aware of those languages, but forgot to consider them for this survey. Thanks!

Also, dammit, Crystal's macros look like an impossible paradise version of Perl 6's macros. I should check them out some more.

**Comment by vendethiel** [comment]
I think they've been mentioned quite a few times over the years on IRC.
Crystal's macro system is basically a pre-proc; that lives in another stage; and doesn't have access to everything else.
Nim macro I remember us discussing (the delaying of type resolution while in expansion etc), but Julia maybe not.

**Comment by masak** [comment]
F# has [code quotations](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/code-quotations), which look a lot like quasiquotes-done-right to me. Even the syntax somehow feels less horrible than par for the course.

**Comment by masak** [comment]
Huh. On page 18 of [the "Resugaring" paper](https://cs.brown.edu/research/pubs/theses/phd/2018/pombrio.justin.pdf) (recommended to me by @jnthn) there is actually a table not entirely unlike what I imagined for this issue. Probably would serve as a very good starting point.

**Comment by masak** [comment]
Agda has "reflection" (a way to turn code into AST objects inside the program itself), as well as quoting/unquoting and macros, as can be seen in [this paper](https://arxiv.org/pdf/2105.10819.pdf). I'm surprised and intrigued, even as I don't see the whole picture. There's something called "the elaborator", and a `TC` monad to maintain the elaborator's context.

**Comment by masak** [comment]
I just found [Axel](https://github.com/axellang/axel/), a Haskell with Lisp syntax and macros, quoting, and quasiquoting+unquotes.

**Comment by masak** [comment]
[Passerine](https://github.com/vrtbl/passerine), which has a respectable pattern macro system that reminds me of Dylan's, but with metasyntax that looks more like Scheme's. Claims to be hygienic.

**Comment by masak** [comment]
[Nemerle](http://nemerle.org/About). Here's a paper: [Meta-programming in Nemerle](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.101.8265&rep=rep1&type=pdf).

**Comment by masak** [comment]
[Wyvern](http://www.cs.cmu.edu/~aldrich/papers/sac15-tsms.pdf).
